### DISCUSSION TITLE
Help! Partial deletion of data when creating the connection of two collections in MongoDB.

### DISCUSSION THREAD

**Discussion by AlexDernov** [neuefische/questions#665] [Web] [answered]
Hi. I have a problem with the database. 

**I have:** 
Two collections in the database (locations and notes to them) . 
**Submitting** the noteForm on the details page **should:** 
**1.** save the note to the collection "notes" (does without problems)
**2.** enter the note's id in the "locations" collection (parameter "notes", data type - array) to link the two collections.

**Problem 1:**
When entering a note id into the "locations" collection, the "height" parameter from the "bild" object (collection "locations") is removed from the collection, and the "width" parameter is moved to the int32 data type instead of int64.

**Problem 2:**
 When adding in the api/[id] page _.populate("notes")_, the detail page stops displaying data from the database at all. 

I can't understand the reasons why this is happening. Search in google and similar codes was unsuccessful.

[My last PR.](https://github.com/AlexDernov/capestone-project-hungryOar/pull/11)

**Comment by Kheirah** [comment]
Try the following and see if it fixes the issue. In your `NotesForm` there's this snippet where you need to remove `locData`:

```js
if (responseNote.ok) {
      const data = await responseNote.json();
      console.log("Undif", data);
      const responseLocation = await fetch(`/api/locations/${locData?._id}`, {
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ //THIS LINE CHANGED (locData removed)
          notes: [...locData?.notes, data.data._id],
        }),
      });
      if (responseLocation.ok) {
        mutate();
      }
    }
```

Then, in your API endpoint `locations([id].js` change another line:
```js
  if (request.method === "PUT") {
    await Location.findByIdAndUpdate(id, request.body); //THIS LINE CHANGED ($set removed)
    response.status(200).json({ message: "Update is successful!" });
    console.log(request.body);
  }
```
In combination, these two changes may fix the issue and no longer remove a property and change the type of another property. See if it works.

**Comment by Kheirah** [comment]
As for the second issue concerning `populate`, try adding the following and inspect the logs. The logs might tell us more about the exact error message.
```js
if (request.method === "GET") {
    const location = await Location.findById(id).populate('notes').then(foundLocation => {
    console.log(foundLocation)
  })
  .catch(err => {
    console.error(err);
  });

    if (!location) {
      return response.status(404).json({ status: "Not found" });
    }
    response.status(200).json(location);
  }
```
Let me know what the logs say and we'll see where to go from there.

Edit:
Fixed my suggested code. `.populate` was missing.

**Reply by AlexDernov** [reply]
Sorry, had to rush to the kids' doctor yesterday, so I'm only replying today.
console.log("FoundLocation", foundLocation) is not displayed.
The console shows the following:
<img width="481" alt="Screenshot 15 08 2023 um 09 05 45 AM" src="https://github.com/neuefische/questions/assets/127208943/4336f6ed-9417-4fa2-9fea-8d15aa99ce82">

**Comment by mbosselmann** [comment, marked as answer]
Hi Alexandra,
I needed some time to figure it out and I also needed to change the code that Ernst suggested. I added the suggestions from Ernst to your project and then removed the `.then()` and `.error()` in `api/locations/`:

```js
  if (request.method === "GET") {
    const location = await Location.findById(id).populate("notes");

    if (!location) {
      return response.status(404).json({ status: "Not found" });
    }
    response.status(200).json(location);
  }

```

And I also got the following error message in my terminal: `Error [MissingSchemaError]: Schema hasn't been registered for model "Notes".`

Therefore, I changed the ref for the notes in your Location model from "Notes" to "Note": `  notes: { type: [Schema.Types.ObjectId], ref: "Note" },`.

After these changes everything worked as I think it should do. :) 

I hope this does help!
